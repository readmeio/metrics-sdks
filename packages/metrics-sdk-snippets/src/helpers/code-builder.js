// https://github.com/readmeio/httpsnippet/blob/19f5557205418ef667bb02c2bd940fa4269adad5/src/helpers/code-builder.js
/* eslint-disable no-param-reassign */
import { format } from 'util';

/**
 * Helper object to format and aggragate lines of code.
 * Lines are aggregated in a `code` array, and need to be joined to obtain a proper code snippet.
 *
 * @class
 *
 * @param {string} indentation Desired indentation character for aggregated lines of code
 * @param {string} join Desired character to join each line of code
 */
class CodeBuilder {
  constructor(indentation, join) {
    this.code = [];
    this.indentation = indentation;
    this.lineJoin = join || '\n';
  }

  /**
   * Add given indentation level to given string and format the string (variadic)
   * @param {number} [indentationLevel=0] - Desired level of indentation for this line
   * @param {string} line - Line of code. Can contain formatting placeholders
   * @param {...anyobject} - Parameter to bind to `line`'s formatting placeholders
   * @return {string}
   *
   * @example
   *   var builder = CodeBuilder('\t')
   *
   *   builder.buildLine('console.log("hello world")')
   *   // returns: 'console.log("hello world")'
   *
   *   builder.buildLine(2, 'console.log("hello world")')
   *   // returns: 'console.log("\t\thello world")'
   *
   *   builder.buildLine(2, 'console.log("%s %s")', 'hello', 'world')
   *   // returns: 'console.log("\t\thello world")'
   */
  buildLine(indentationLevel, line) {
    let lineIndentation = '';
    let slice = 2;
    if (Object.prototype.toString.call(indentationLevel) === '[object String]') {
      slice = 1;
      line = indentationLevel;
      indentationLevel = 0;
    } else if (indentationLevel === null) {
      return null;
    }

    while (indentationLevel) {
      lineIndentation += this.indentation;
      // eslint-disable-next-line no-plusplus
      indentationLevel--;
    }

    // eslint-disable-next-line prefer-rest-params
    const data = Array.prototype.slice.call(arguments, slice, arguments.length);
    data.unshift(lineIndentation + line);

    return format.apply(this, data);
  }

  /**
   * Invoke buildLine() and add the line at the top of current lines
   * @param {number} [indentationLevel=0] Desired level of indentation for this line
   * @param {string} line Line of code
   * @return {this}
   */
  unshift(...args) {
    // eslint-disable-next-line prefer-spread
    this.code.unshift(this.buildLine.apply(this, args));

    return this;
  }

  /**
   * Invoke buildLine() and add the line at the bottom of current lines
   * @param {number} [indentationLevel=0] Desired level of indentation for this line
   * @param {string} line Line of code
   * @return {this}
   */
  push(...args) {
    // eslint-disable-next-line prefer-spread
    this.code.push(this.buildLine.apply(this, args));

    return this;
  }

  /**
   * Add an empty line at the end of current lines
   * @return {this}
   */
  blank() {
    this.code.push(null);

    return this;
  }

  /**
   * Concatenate all current lines using the given lineJoin
   * @return {string}
   */
  join() {
    return this.code.join(this.lineJoin);
  }
}

export default CodeBuilder;
