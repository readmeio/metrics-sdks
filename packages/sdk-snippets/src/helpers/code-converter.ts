import { basename } from 'node:path';
import process from 'process';

async function getStdin() {
  if (process.stdin.isTTY) return null;

  const result = [];
  // eslint-disable-next-line no-restricted-syntax
  for await (const chunk of process.stdin) {
    result.push(chunk);
  }

  return result.join('');
}

function processLine(line) {
  if (line === '') return '\nblank()\n';
  const indentation = line.search(/\S|$/) / 2;
  const quote = line.match(/'/) ? '"' : "'";
  return `push(${quote}${line.trimLeft()}${quote}${indentation !== 0 ? `, ${indentation}` : ''});`;
}

export default function main(input) {
  return input
    .toString()
    .split('\n')
    .map(line => processLine(line))
    .join('\n');
}

if (basename(process.argv[1]) === 'code-converter.ts') {
  (async () => {
    const stdin = await getStdin();
    if (!stdin) {
      // eslint-disable-next-line no-console
      console.error('You must pipe something into this file');
      return process.exit(1);
    }

    // eslint-disable-next-line no-console
    return console.log(main(stdin));
  })();
}
