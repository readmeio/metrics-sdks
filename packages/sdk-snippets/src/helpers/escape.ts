export interface EscapeOptions {
  /**
   * The delimiter that will be used to wrap the string (and so must be escaped
   * when used within the string).
   * Defaults to "
   */
  delimiter?: string;

  /**
   * The char to use to escape the delimiter and other special characters.
   * Defaults to \
   */
  escapeChar?: string;

  /**
   * Whether newlines (\n and \r) should be escaped within the string.
   * Defaults to true.
   */
  escapeNewlines?: boolean;
}

/**
 * Escape characters within a value to make it safe to insert directly into a
 * snippet. Takes options which define the escape requirements.
 *
 * This is closely based on the JSON-stringify string serialization algorithm,
 * but generalized for other string delimiters (e.g. " or ') and different escape
 * characters (e.g. Powershell uses `)
 *
 * See https://tc39.es/ecma262/multipage/structured-data.html#sec-quotejsonstring
 * for the complete original algorithm.
 */
export function escapeString(rawValue: any, options: EscapeOptions = {}) {
  const { delimiter = '"', escapeChar = '\\', escapeNewlines = true } = options;

  const stringValue = rawValue.toString();

  return [...stringValue]
    .map(c => {
      if (c === '\b') {
        return `${escapeChar}b`;
      } else if (c === '\t') {
        return `${escapeChar}t`;
      } else if (c === '\n') {
        if (escapeNewlines) {
          return `${escapeChar}n`;
        }
        return c; // Don't just continue, or this is caught by < \u0020
      } else if (c === '\f') {
        return `${escapeChar}f`;
      } else if (c === '\r') {
        if (escapeNewlines) {
          return `${escapeChar}r`;
        }
        return c; // Don't just continue, or this is caught by < \u0020
      } else if (c === escapeChar) {
        return escapeChar + escapeChar;
      } else if (c === delimiter) {
        return escapeChar + delimiter;
      } else if (c < '\u0020' || c > '\u007E') {
        // Delegate the trickier non-ASCII cases to the normal algorithm. Some of these
        // are escaped as \uXXXX, whilst others are represented literally. Since we're
        // using this primarily for header values that are generally (though not 100%
        // strictly?) ASCII-only, this should almost never happen.
        return JSON.stringify(c).slice(1, -1);
      }
      return c;
    })
    .join('');
}

/**
 * Make a string value safe to insert literally into a snippet within single quotes,
 * by escaping problematic characters, including single quotes inside the string,
 * backslashes, newlines, and other special characters.
 *
 * If value is not a string, it will be stringified with .toString() first.
 */
export const escapeForSingleQuotes = (value: any) => escapeString(value, { delimiter: "'" });

/**
 * Make a string value safe to insert literally into a snippet within double quotes,
 * by escaping problematic characters, including double quotes inside the string,
 * backslashes, newlines, and other special characters.
 *
 * If value is not a string, it will be stringified with .toString() first.
 */
export const escapeForDoubleQuotes = (value: any) => escapeString(value, { delimiter: '"' });

function quotedString(str: string, doubleQuotes: boolean) {
  if (doubleQuotes) {
    return `"${str}"`;
  }

  return `'${str}'`;
}

/**
 * Safely escape and prepare a string to be used as an object key.
 *
 */
export function escapeForObjectKey(str: string, doubleQuotes = false) {
  const escaped = doubleQuotes ? escapeForDoubleQuotes(str) : escapeForSingleQuotes(str);

  // If this key doesn't start with an alpha character then we should always wrap it in quotes and
  // then run it through some escaping logic to make sure that it doesn't contain anything that
  // might break out of the quotes.
  if (!/^([A-Za-z]{1})/.test(str)) {
    return quotedString(escaped, doubleQuotes);
  } else if (escaped === str) {
    // If our key doesn't have any characters that might break out of quotes and requring escaping
    // but might have any non-alphanumeric characters in it that might not be able to be used as
    // a JS property key then we should wrap the key in quotes.
    if (/([^A-Za-z_0-9])/.test(str)) {
      return quotedString(str, doubleQuotes);
    }

    // If our key doesn't need to be escaped then we don't need to wrap it in quotes.
    return str;
  }

  return quotedString(str, doubleQuotes);
}
