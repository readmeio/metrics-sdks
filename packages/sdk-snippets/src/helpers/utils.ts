import type { ClientInfo, TargetId, TargetInfo } from '../targets/targets';

import { targets } from '../targets/targets';

export interface AvailableTarget extends TargetInfo {
  clients: ClientInfo[];
}

export const availableWebhookTargets = () =>
  Object.keys(targets).map<AvailableTarget>(targetId => ({
    ...targets[targetId as TargetId].info,
    clients: Object.keys(targets[targetId as TargetId].services.webhooks).map(
      clientId => targets[targetId as TargetId].services.webhooks[clientId].info,
    ),
  }));

export const availableServerTargets = () =>
  Object.keys(targets).map<AvailableTarget>(targetId => ({
    ...targets[targetId as TargetId].info,
    clients: Object.keys(targets[targetId as TargetId].services.server).map(
      clientId => targets[targetId as TargetId].services.server[clientId].info,
    ),
  }));

export const extname = (targetId: TargetId) => targets[targetId]?.info.extname || '';
