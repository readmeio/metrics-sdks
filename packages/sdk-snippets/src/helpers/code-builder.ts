import type { CodeBuilderOptions } from '@readme/httpsnippet/dist/helpers/code-builder';

import { CodeBuilder as HTTPSnippetCodeBuilder } from '@readme/httpsnippet/dist/helpers/code-builder';

export type { CodeBuilderOptions };

export class CodeBuilder extends HTTPSnippetCodeBuilder {
  sections: {
    payload?: {
      end: number;
      start: number;
    };
    verification?: {
      end: number;
      start: number;
    };
  } = {};

  variables: {
    security?: Record<string, { line: number }>;
    server?: Record<string, { line: number }>;
  } = {};

  startSection = (section: 'payload' | 'verification') => {
    this.sections[section] = {
      start: this.code.length + 1,
      end: 0,
    };
  };

  endSection = (section: 'payload' | 'verification') => {
    this.sections[section].end = this.code.length;
  };

  ranges = () => {
    return {
      sections: this.sections,
      variables: this.variables,
    };
  };

  /**
   * Push a dynamic piece of code to the snippet and record where in the snippet it was added.
   *
   */
  pushVariable = (
    line: string,
    opts: {
      indentationLevel?: number;
      name: string;
      type: 'security' | 'server';
    },
  ) => {
    this.push(line, opts.indentationLevel);

    // Record where in the snippet this variable is located.
    if (!this.variables[opts.type]) {
      this.variables[opts.type] = {};
    }

    this.variables[opts.type][opts.name] = {
      line: this.code.length,
    };
  };
}
