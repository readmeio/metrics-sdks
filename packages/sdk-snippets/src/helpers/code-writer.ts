import CodeBlockWriter from 'code-block-writer';

import { escapeForDoubleQuotes, escapeForSingleQuotes } from './escape';

export class CodeWriter extends CodeBlockWriter {
  private ranges: {
    security?: Record<string, { line: number }>;
    server?: Record<string, { line: number }>;
  } = {};

  private _texts: string[] = [];

  getRanges = () => {
    return this.ranges;
  };

  recordRange = (type: 'security' | 'server', name: string) => {
    if (!this.ranges[type]) {
      this.ranges[type] = {};
    }

    this.ranges[type][name] = {
      // eslint-disable-next-line no-underscore-dangle
      line: this._texts.join('').trim().split(this.getOptions().newLine).length,
    };

    return this;
  };

  writeEscapedString = (str: string) => {
    if (this.getOptions().useSingleQuote) {
      return this.write(`'${escapeForSingleQuotes(str)}'`);
    }

    return this.write(`"${escapeForDoubleQuotes(str)}"`);
  };

  /**
   * Safely escape and write a string as an object key.
   *
   */
  writeObjectKey = (str: string) => {
    const { useSingleQuote } = this.getOptions();

    const escaped = useSingleQuote ? escapeForSingleQuotes(str) : escapeForDoubleQuotes(str);

    // If this key doesn't start with an alpha character then we should always wrap it in quotes and
    // then run it through some escaping logic to make sure that it doesn't contain anything that
    // might break out of the quotes.
    if (!/^([A-Za-z]{1})/.test(str)) {
      return this.write(useSingleQuote ? `'${escaped}'` : `"${escaped}"`);
    } else if (escaped === str) {
      // If our key doesn't have any characters that might break out of quotes and requring escaping
      // but might have any non-alphanumeric characters in it that might not be able to be used as
      // a JS property key then we should wrap the key in quotes.
      if (/([^A-Za-z_0-9])/.test(str)) {
        return this.write(useSingleQuote ? `'${str}'` : `"${str}"`);
      }

      // If our key doesn't need to be escaped then we don't need to wrap it in quotes.
      return this.write(str);
    }

    return this.write(useSingleQuote ? `'${str}'` : `"${str}"`);
  };
}
