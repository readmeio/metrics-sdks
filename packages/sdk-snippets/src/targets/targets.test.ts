import type { ClientId, SnippetType, TargetId } from './targets';
import type { Variables } from '..';

import { readdirSync, readFileSync, writeFileSync } from 'fs';
import path from 'path';

import { describe, it, expect } from 'vitest';

import { MetricsSDKSnippet } from '..';
import { availableWebhookTargets, extname } from '../helpers/utils';

const expectedBasePath = ['src', 'fixtures', 'webhooks'];

const inputFileNames = readdirSync(path.join(...expectedBasePath), 'utf-8');

const fixtures: [string, Variables][] = inputFileNames.map(inputFileName => [
  inputFileName.replace(path.extname(inputFileName), ''),
  // eslint-disable-next-line import/no-dynamic-require, global-require
  require(path.resolve(...expectedBasePath, inputFileName)),
]);

/** useful for debuggin, only run a particular set of targets */
const targetFilter: TargetId[] = [
  // put your targetId:
  // 'node',
];

/** useful for debuggin, only run a particular set of targets */
const clientFilter: ClientId[] = [
  // put your clientId here:
  // 'express',
];

/** useful for debuggin, only run a particular set of fixtures */
const fixtureFilter: string[] = [
  // put the name of the fixture file you want to isolate (excluding `.js`):
  // 'security-variables',
];

/**
 * This is useful when you want to make a change and overwrite it for every fixture.
 * Basically a snapshot reset.
 *
 * Switch to `true` in debug mode to put into effect.
 */
const OVERWRITE_EVERYTHING = Boolean(process.env.OVERWRITE_EVERYTHING) || false;

const testFilter =
  <T>(property: keyof T, list: T[keyof T][]) =>
  (item: T) =>
    list.length > 0 ? list.includes(item[property]) : true;

describe('webhooks', function () {
  // eslint-disable-next-line vitest/require-hook
  availableWebhookTargets()
    .filter(testFilter('key', targetFilter))
    .forEach(({ key: targetId, title, clients }) => {
      const snippetType: SnippetType = 'webhooks';

      describe(`${title} snippet generation`, function () {
        // eslint-disable-next-line vitest/require-hook
        clients.filter(testFilter('key', clientFilter)).forEach(({ key: clientId }) => {
          describe(clientId, function () {
            // eslint-disable-next-line vitest/require-hook
            fixtures.filter(testFilter(0, fixtureFilter)).forEach(([fixture, variables]) => {
              const fixturePath = path.join(
                'src',
                'targets',
                targetId,
                clientId,
                snippetType,
                'fixtures',
                `${fixture}${extname(targetId)}`,
              );

              let expectedOutput: string;
              try {
                expectedOutput = readFileSync(fixturePath).toString();
              } catch (err) {
                if (OVERWRITE_EVERYTHING) {
                  writeFileSync(fixturePath, '');
                  return;
                }

                throw new Error(
                  `Missing a ${snippetType} test file for ${targetId}:${clientId} for the ${fixture} fixture.\nExpected to find the output fixture: \`/${fixturePath}\``,
                );
              }

              it(`request should match fixture for "${fixture}"`, function () {
                const { convert } = new MetricsSDKSnippet(variables);
                const result = convert(snippetType, targetId, clientId);

                if (OVERWRITE_EVERYTHING && result) {
                  writeFileSync(fixturePath, String(result.snippet));
                  return;
                }

                if (!result) {
                  throw new Error(`Generated ${fixture} snippet for ${clientId} was \`false\``);
                }

                // This test is to make sure that our generated snippets actually do any variable
                // outputting vs being static.
                if (fixture !== 'empty') {
                  expect(Object.keys(result.ranges).length).to.be.greaterThan(0);
                }

                expect(result.ranges).toMatchSnapshot();
                expect(result.snippet).to.deep.equal(expectedOutput);

                // This is making sure that there is an actual secret in the generated output.
                expect(result.snippet).to.match(/my-readme-secret/);
              });
            });

            if (clientId === 'aws') {
              describe('AWS snippet generation with automatic creation of new API keys', function () {
                // eslint-disable-next-line vitest/require-hook
                fixtures.filter(testFilter(0, fixtureFilter)).forEach(([fixture, variables]) => {
                  const createKeys = true;
                  const suffix = createKeys ? '-create' : '';
                  const fixturePath = path.join(
                    'src',
                    'targets',
                    targetId,
                    clientId,
                    snippetType,
                    'fixtures',
                    `${fixture}${suffix}${extname(targetId)}`,
                  );

                  let expectedOutput: string;
                  try {
                    expectedOutput = readFileSync(fixturePath).toString();
                  } catch (err) {
                    if (OVERWRITE_EVERYTHING) {
                      writeFileSync(fixturePath, '');
                      return;
                    }
                    throw new Error(
                      `Missing a ${snippetType} test file for ${targetId}:${clientId} for the ${fixture} fixture.\nExpected to find the output fixture: \`/${fixturePath}\``,
                    );
                  }

                  const { convert } = new MetricsSDKSnippet(variables);
                  const result = convert(snippetType, targetId, clientId, { createKeys });

                  if (OVERWRITE_EVERYTHING && result) {
                    writeFileSync(fixturePath, String(result.snippet));
                    return;
                  }

                  it(`${clientId} request should match fixture for "${fixture}${suffix}"`, function () {
                    if (!result) {
                      throw new Error(`Generated ${fixture} snippet for ${clientId} was \`false\``);
                    }

                    // This test is to make sure that our generated snippets actually do any
                    // variable outputting vs being static.
                    if (fixture !== 'empty') {
                      expect(Object.keys(result.ranges).length).to.be.greaterThan(0);
                    }

                    expect(result.ranges).toMatchSnapshot();
                    expect(result.snippet).to.deep.equal(expectedOutput);

                    // This is making sure that there is an actual secret in the generated output.
                    expect(result.snippet).to.match(/my-readme-secret/);
                  });
                });
              });
            }
          });
        });
      });
    });
});
