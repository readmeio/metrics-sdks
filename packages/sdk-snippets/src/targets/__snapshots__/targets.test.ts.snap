// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webhooks Node.js snippet generation express request should match fixture for "empty" 1`] = `Object {}`;

exports[`webhooks Node.js snippet generation express request should match fixture for "empty-default" 1`] = `
Object {
  "security": Object {
    "petstore_auth": Object {
      "line": 29,
    },
  },
  "server": Object {
    "name": Object {
      "line": 26,
    },
  },
}
`;

exports[`webhooks Node.js snippet generation express request should match fixture for "security-and-server-variables" 1`] = `
Object {
  "security": Object {
    "basic_auth": Object {
      "line": 31,
    },
    "petstore_auth": Object {
      "line": 30,
    },
  },
  "server": Object {
    "name": Object {
      "line": 26,
    },
    "port": Object {
      "line": 27,
    },
  },
}
`;

exports[`webhooks Node.js snippet generation express request should match fixture for "security-types" 1`] = `
Object {
  "security": Object {
    "api_key": Object {
      "line": 26,
    },
    "http_basic": Object {
      "line": 27,
    },
    "http_bearer": Object {
      "line": 28,
    },
    "oauth2": Object {
      "line": 29,
    },
  },
}
`;

exports[`webhooks Node.js snippet generation express request should match fixture for "security-variables" 1`] = `
Object {
  "security": Object {
    "basic_auth": Object {
      "line": 27,
    },
    "petstore_auth": Object {
      "line": 26,
    },
  },
}
`;

exports[`webhooks Node.js snippet generation express request should match fixture for "server-variables" 1`] = `
Object {
  "server": Object {
    "name": Object {
      "line": 26,
    },
    "port": Object {
      "line": 27,
    },
  },
}
`;
